README
NAME

    Mojo::S3Object - Mojo interface to S3 objects.

VERSION

    0.02_5

SOURCE REPOSITORY

    http://github.com/keenlinks/Mojo-S3Objects

AUTHOR

    Scott Kiehn <sk.keenlinks@gmail.com>

COPYRIGHT

    Copyright 2018 - Scott Kiehn

LICENSE

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO



Changes
Revision history for Perl module Mojo-S3Object

0.02_5 2018-01-25
  - Added ability to add object user-defined metadata headers to the "put" method.
  - Bug fixed "update" method as it uses the "put" method.

0.02_4 2018-01-25
  - Added ability to add object metadata headers to the "put" method.

0.02_3 2017-10-17
  - Changed private method to public (sign_the_string).

0.02_2 2017-05-18
  - Changed from a specific Mojolicious plugin
    to a Mojo namespaced module (can be used 
    outside a Mojolicious app)
  - Renamed to S3Object as it does not
    handle bucket functions, just the
    objects.
  - Renamed methods (removed "_file" from each).
  - Created distinct "upload" function (uses the "get"
    function, but passes in a Mojo::Upload object.

0.02_1 2017-03-03
  - Created header private sub
  - Renamed helper move_file to copy_file and removed delete_file from copy_file.
    S3 api does not have a move or delete with a copy.
  - Refined non-blocking code.

0.01_2 2017-02-28
  - Add some non-blocking code

0.01_1 2017-02-27
  - Initial dev release


lib/Mojolicious/Plugin/S3Request.pm
package Mojolicious::Plugin::S3Request;

use Mojo::Base 'Mojolicious::Plugin';
use Digest::SHA qw(hmac_sha256 hmac_sha256_hex sha256_hex);
use Mojo::Date;
use Mojo::IOLoop;
use Mojo::JSON qw(encode_json);
use Mojo::Util qw(monkey_patch trim);

monkey_patch 'Mojo::Date', to_ymd => sub {
	my @gmtime = gmtime shift->epoch;
	sprintf '%04d%02d%02d', $gmtime[5] + 1900, $gmtime[4] + 1, $gmtime[3];
};

monkey_patch 'Mojo::Date', to_ymdhms => sub {
	my @gmtime = gmtime shift->epoch;
	sprintf '%04d%02d%02dT%02d%02d%02dZ', $gmtime[5] + 1900, $gmtime[4] + 1, $gmtime[3], $gmtime[2], $gmtime[1], $gmtime[0];
};

our $VERSION = '0.02_1';

has ioloop => sub { Mojo::IOLoop->new };
has aws_auth_alg => 'AWS4-HMAC-SHA256';
has aws_host => 's3.amazonaws.com';
has aws_request => 'aws4_request';
has aws_service => 's3';

sub register {
	my ( $self, $app, $conf ) = @_;

	$self->{conf} = $conf || {};

	# Required
	$self->{conf}->{access_key_id} || die 'Access key id value is mandatory.';
	$self->{conf}->{access_key}    || die 'Access key value is mandatory.';
	$self->{conf}->{bucket}        || die 'A bucket value is mandatory.';

	# Defaults if not provided
	$self->{conf}->{protocol} ||= 'https://';
	$self->{conf}->{region}   ||= 'us-east-1';

	$app->helper( 's3.bucket_url' => sub { $self->{conf}->{protocol} . $self->_bucket_path . '/' });

	$app->helper( 's3.copy_file' => sub {
		my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
		my ( $c, $former_file, $file ) = @_;

		my $headers = $self->_headers( 'PUT', $file, { 'x-amz-copy-source' => $self->{conf}->{bucket} . '/' . $former_file });

		return $c->ua->put( $self->_object_url( $file ) => $headers )->res->code == 200 ? 'ok' : undef
			unless $cb;

		$self->ioloop->delay(
			sub {
				$c->ua->put( $self->_object_url( $file ) => $headers, shift->begin );
			},
			sub {
				return $self->$cb( pop->res->code == 200 ? 'ok' : undef );
			}
		)->wait;
	});

	$app->helper( 's3.delete_file' => sub {
		my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
		my ( $c, $file ) = @_;

		my $headers = $self->_headers( 'DELETE', $file );

		return $c->ua->delete( $self->_object_url( $file ) => $headers )->res->code == 204 ? 'ok' : undef
			unless $cb;

		$self->ioloop->delay(
			sub {
				$c->ua->delete( $self->_object_url( $file ) => $headers, shift->begin );
			},
			sub {
				return $self->$cb( pop->res->code == 204 ? 'ok' : undef );
			}
		)->wait;
	});

	$app->helper( 's3.get_file' => sub {
		my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
		my ( $c, $file ) = @_;

		my $headers = $self->_headers( 'GET', $file );

		unless ( $cb ) {
			my $tx = $c->ua->get( $self->_object_url( $file ) => $headers );
			return $tx->res->code == 200 ? $tx->res : undef;
		}

		$self->ioloop->delay(
			sub {
				$c->ua->get( $self->_object_url( $file ) => $headers, shift->begin );
			},
			sub {
				my $tx = pop;
				return $self->$cb( $tx->res->code == 200 ? $tx->res : undef );
			}
		)->wait;
	});

	$app->helper( 's3.put_file' => sub {
		my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
		my ( $c, $file, $upload ) = @_;
		my $content = $upload->asset->slurp;

		my $headers = $self->_headers( 'PUT', $file, {
			'content-length' => $upload->size,
			'content-type' => $upload->headers->content_type,
			'x-amz-content-sha256' => sha256_hex( $content )
		});

		return $c->ua->put( $self->_object_url( $file ) => $headers => $content )->res->code == 200 ? 'ok' : undef
			unless $cb;

		$self->ioloop->delay(
			sub {
				$c->ua->put( $self->_object_url( $file ) => $headers => $content, shift->begin );
			},
			sub {
				return $self->$cb( pop->res->code == 200 ? 'ok' : undef );
			}
		)->wait;
	});

	$app->helper( 's3.browser_policy' => sub {
		my ( $c, $params ) = @_;
		my $date = Mojo::Date->new;
		my $credential = $self->_credential( $date );

		my $policy = encode_json({
			expiration => Mojo::Date->new( time + 900 )->to_datetime,
			conditions => [
				{bucket => $self->{conf}->{bucket}},
				{'x-amz-algorithm' => $self->aws_auth_alg},
				{'x-amz-credential' => $credential},
				{'x-amz-date' => $date->to_ymdhms},
				{acl => 'public-read'},
				{key => $params->{filename}},
				{'content-type' => $params->{filetype}},
				['content-length-range', 0, $params->{filesize}],
				{success_action_status => '200'}
			]
		});

		my $base64_policy = $c->b( $policy )->b64_encode('');

		return {
			'x-amz-algorithm' => $self->aws_auth_alg,
			'x-amz-credential' => $credential,
			'x-amz-date' => $date->to_ymdhms,
			acl => 'public-read',
			key => $params->{filename},
			'content-type' => $params->{filetype},
			success_action_status => '200',
			policy => $base64_policy,
			'x-amz-signature' => $self->_sign_the_string( $base64_policy ),
		};
	});
}

sub _auth_header {
	my ( $self, $method, $date, $file, $raw_headers ) = @_;

	my $headers = {};
	map { $headers->{lc trim $_} = $raw_headers->{$_} } keys %$raw_headers;

	my $string_to_sign = $self->_string_to_sign( $self->_canonical_request( $method, $file, $headers ), $date );

	my $signature = $self->_sign_the_string( $string_to_sign, $self->{conf}->{access_key}, $date->to_ymd );

	my $auth = $self->aws_auth_alg;
	$auth .= ' Credential=' . $self->_credential( $date ) . ',';
	$auth .= 'SignedHeaders=' . ( join ';', sort keys %$headers ) . ',';
	$auth .= 'Signature=' . $signature;
}

sub _bucket_path {
	$_[0]->{conf}->{bucket} . '.' . $_[0]->aws_host;
}

sub _canonical_request {
	my ( $self, $method, $file, $headers ) = @_;
	my $canonical_request = $method . "\n";
	$canonical_request .=  '/' . $file . "\n";
	$canonical_request .= "\n";
	$canonical_request .= $_ . ':' . $headers->{$_} . "\n" for sort keys %$headers;
	$canonical_request .= "\n" . ( join ';', sort keys %$headers ) . "\n";
	$canonical_request .= $headers->{'x-amz-content-sha256'};
}

sub _credential {
	$_[0]->{conf}->{access_key_id} . '/' . $_[1]->to_ymd . '/' . $_[0]->{conf}->{region} . '/' . $_[0]->aws_service . '/' . $_[0]->aws_request;
}

sub _headers {
	my ( $self, $method, $file, $headers ) = @_;
	my $date = Mojo::Date->new;
	$headers->{date} = $date->to_string;
	$headers->{host} = $self->_bucket_path;
	$headers->{'x-amz-content-sha256'} = sha256_hex( 'UNSIGNED-PAYLOAD' ) unless defined $headers->{'x-amz-content-sha256'};
	$headers->{authorization} = $self->_auth_header( $method, $date, $file, $headers );
	return $headers;
}

sub _object_url { $_[0]->{conf}->{protocol} . $_[0]->_bucket_path . '/' . $_[1] }

sub _signing_key {
	my $self = shift;
	hmac_sha256( $self->aws_request,
		hmac_sha256( $self->aws_service,
			hmac_sha256( $self->{conf}->{region},
				hmac_sha256( Mojo::Date->new->to_ymd, 'AWS4' . $self->{conf}->{access_key} )
			)
		)
	);
}

sub _sign_the_string { hmac_sha256_hex( $_[1], $_[0]->_signing_key ) }

sub _string_to_sign {
	my ( $self, $canonical_request, $date ) = @_;
	my $string_to_sign = $self->aws_auth_alg . "\n";
	$string_to_sign .= $date->to_string . "\n";
	$string_to_sign .= $date->to_ymd . '/' . $self->{conf}->{region} . '/' . $self->aws_service . '/' . $self->aws_request . "\n";
	$string_to_sign .= sha256_hex( $canonical_request );
}

1;
__END__

=encoding utf-8

=head1 NAME

Mojolicious::Plugin::S3Request - Mojolicious Plugin for basic S3 AWS4 requests.

=head1 VERSION

0.02_1

=head1 SOURCE REPOSITORY

L<http://github.com/keenlinks/Mojolicious-Plugin-S3Request>

=head1 AUTHOR

Scott Kiehn E<lt>sk.keenlinks@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2017 - Scott Kiehn

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut


lib/Mojo/S3Object.pm
package Mojo::S3Object;

use Mojo::Base -base;
use Carp;
use Digest::SHA qw(hmac_sha256 hmac_sha256_hex sha256_hex);
use Mojo::Date;
use Mojo::IOLoop;
use Mojo::JSON qw(encode_json);
use Mojo::UserAgent;
use Mojo::Util qw(b64_encode monkey_patch slugify);

our $VERSION = '0.02_5';

# Required
has access_key => sub { croak 'Parameter "access_key" is mandatory in the constructor.' };
has access_key_id => sub { croak 'Parameter "access_key_id" is mandatory in the constructor.' };
has bucket => sub { croak 'Parameter "bucket" is mandatory in the constructor.' };

# Defaults if not provided
has protocol => sub { 'https://' };
has region => sub { 'us-east-1' };

# Defaults
has aws_auth_alg => 'AWS4-HMAC-SHA256';
has aws_host => 's3.amazonaws.com';
has aws_request => 'aws4_request';
has aws_service => 's3';
has policy_exporation => 900; # 15 minutes

has bucket_path => sub { $_[0]->bucket . '.' . $_[0]->aws_host };
has bucket_url => sub { $_[0]->protocol . $_[0]->bucket_path };

has ua => sub { state $ua = Mojo::UserAgent->new };

monkey_patch 'Mojo::Date', to_ymd => sub {
	my @gmtime = gmtime shift->epoch;
	sprintf '%04d%02d%02d', $gmtime[5] + 1900, $gmtime[4] + 1, $gmtime[3];
};

monkey_patch 'Mojo::Date', to_ymdhms => sub {
	my @gmtime = gmtime shift->epoch;
	sprintf '%04d%02d%02dT%02d%02d%02dZ', $gmtime[5] + 1900, $gmtime[4] + 1, $gmtime[3], $gmtime[2], $gmtime[1], $gmtime[0];
};

sub browser_policy {
	my ( $self, $params ) = @_;
	my $date = Mojo::Date->new;
	my $credential = $self->_credential( $date );

	my $policy = encode_json({
		expiration => Mojo::Date->new( time + $self->policy_exporation )->to_datetime,
		conditions => [
			{bucket => $self->bucket},
			{'x-amz-algorithm' => $self->aws_auth_alg},
			{'x-amz-credential' => $credential},
			{'x-amz-date' => $date->to_ymdhms},
			{acl => 'public-read'},
			{key => $params->{filename}},
			{'content-type' => $params->{filetype}},
			['content-length-range', 0, $params->{filesize}],
			{success_action_status => '200'}
		]
	});

	my $base64_policy = b64_encode $policy, '';

	return {
		'x-amz-algorithm' => $self->aws_auth_alg,
		'x-amz-credential' => $credential,
		'x-amz-date' => $date->to_ymdhms,
		acl => 'public-read',
		key => $params->{filename},
		'content-type' => $params->{filetype},
		success_action_status => '200',
		policy => $base64_policy,
		'x-amz-signature' => $self->sign_the_string( $base64_policy ),
	};
}

sub copy {
	my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
	my ( $self, $former_object_name, $s3_object_name ) = @_;

	my $headers = $self->_headers( 'PUT', $s3_object_name, { 'x-amz-copy-source' => $self->bucket . '/' . $former_object_name });

	return $self->ua->put( $self->_object_url( $s3_object_name ) => $headers )->res->code == 200 ? 'ok' : undef
		unless $cb;

	$self->_ioloop->delay(
		sub {
			$self->ua->put( $self->_object_url( $s3_object_name ) => $headers, shift->begin );
		},
		sub {
			return $self->$cb( pop->res->code == 200 ? 'ok' : undef );
		}
	)->wait;
}

sub delete {
	my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
	my ( $self, $s3_object_name ) = @_;

	my $headers = $self->_headers( 'DELETE', $s3_object_name );

	return $self->ua->delete( $self->_object_url( $s3_object_name ) => $headers )->res->code == 204 ? 'ok' : undef
		unless $cb;

	$self->_ioloop->delay(
		sub {
			$self->ua->delete( $self->_object_url( $s3_object_name ) => $headers, shift->begin );
		},
		sub {
			return $self->$cb( pop->res->code == 204 ? 'ok' : undef );
		}
	)->wait;
}

sub get {
	my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
	my ( $self, $s3_object_name ) = @_;

	my $headers = $self->_headers( 'GET', $s3_object_name );

	unless ( $cb ) {
		my $tx = $self->ua->get( $self->_object_url( $s3_object_name ) => $headers );
		return $tx->res->code == 200 ? $tx->res : undef;
	}

	$self->_ioloop->delay(
		sub {
			$self->ua->get( $self->_object_url( $s3_object_name ) => $headers, shift->begin );
		},
		sub {
			my $tx = pop;
			return $self->$cb( $tx->res->code == 200 ? $tx->res : undef );
		}
	)->wait;
}

sub put {
	my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
	my ( $self, $s3_object_name, $asset, $meta_data, $user_defined ) = @_;

	croak 'You want to use the "upload" method. "Put" method does not use Mojo::Upload.' if ref $asset eq 'Mojo::Upload';

	my $content = $asset->slurp;
	my $prep = { 'content-length' => $asset->size, 'x-amz-content-sha256' => sha256_hex( $content ) };

	if ( defined $meta_data ) {
		if ( ref $meta_data eq 'HASH' ) {
			map { $prep->{ slugify $_ } = $meta_data->{$_} } keys %$meta_data;
		} else {
			carp 'Metadata not applied. Please pass metadata as a hash reference.';
		}
	}

	if ( defined $user_defined ) {
		if ( ref $user_defined eq 'HASH' ) {
			map { $prep->{'x-amz-meta-' . ( slugify $_ )} = $user_defined->{$_} } keys %$user_defined;
		} else {
			carp 'User-defined not applied. Please pass data as a hash reference.';
		}
	}

	my $headers = $self->_headers( 'PUT', $s3_object_name, $prep );

	return $self->ua->put( $self->_object_url( $s3_object_name ) => $headers => $content )->res->code == 200 ? 'ok' : 'oops'
		unless $cb;

	$self->_ioloop->delay(
		sub {
			$self->ua->put( $self->_object_url( $s3_object_name ) => $headers => $content, shift->begin );
		},
		sub {
			my $tx = pop;
			return $self->$cb( $tx->res->code == 200 ? 'ok' : undef );
		}
	)->wait;
}

sub sign_the_string { hmac_sha256_hex( $_[1], $_[0]->_signing_key ) }

sub upload {
	my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
	my ( $self, $s3_object_name, $upload, $meta_data, $user_defined ) = @_;

	croak 'You want to use the "put" method. "Upload" method for use with Mojo::Upload.' unless ref $upload eq 'Mojo::Upload';

	$meta_data->{'content-type'} = $upload->headers->content_type;

	return $self->put( $s3_object_name, $upload->asset, $meta_data, $user_defined, $cb );
}

sub _auth_header {
	my ( $self, $method, $date, $s3_object_name, $raw_headers ) = @_;

	my $headers = {};
	map { $headers->{ slugify $_ } = $raw_headers->{$_} } keys %$raw_headers;

	my $string_to_sign = $self->_string_to_sign( $self->_canonical_request( $method, $s3_object_name, $headers ), $date );

	my $signature = $self->sign_the_string( $string_to_sign, $self->access_key, $date->to_ymd );

	my $auth = $self->aws_auth_alg;
	$auth .= ' Credential=' . $self->_credential( $date ) . ',';
	$auth .= 'SignedHeaders=' . ( join ';', sort keys %$headers ) . ',';
	$auth .= 'Signature=' . $signature;
}

sub _canonical_request {
	my ( $self, $method, $s3_object_name, $headers ) = @_;
	my $canonical_request = $method . "\n";
	$canonical_request .=  '/' . $s3_object_name . "\n";
	$canonical_request .= "\n";
	$canonical_request .= $_ . ':' . $headers->{$_} . "\n" for sort keys %$headers;
	$canonical_request .= "\n" . ( join ';', sort keys %$headers ) . "\n";
	$canonical_request .= $headers->{'x-amz-content-sha256'};
}

sub _credential {
	$_[0]->access_key_id . '/' . $_[1]->to_ymd . '/' . $_[0]->region . '/' . $_[0]->aws_service . '/' . $_[0]->aws_request;
}

sub _headers {
	my ( $self, $method, $s3_object_name, $headers ) = @_;
	my $date = Mojo::Date->new;
	$headers->{date} = $date->to_string;
	$headers->{host} = $self->bucket_path;
	$headers->{'x-amz-content-sha256'} = sha256_hex( 'UNSIGNED-PAYLOAD' ) unless defined $headers->{'x-amz-content-sha256'};
	$headers->{authorization} = $self->_auth_header( $method, $date, $s3_object_name, $headers );
	return $headers;
}

sub _ioloop { Mojo::IOLoop->singleton }

sub _object_url { $_[0]->bucket_url . '/' . $_[1] }

sub _signing_key {
	my $self = shift;
	hmac_sha256( $self->aws_request,
		hmac_sha256( $self->aws_service,
			hmac_sha256( $self->region,
				hmac_sha256( Mojo::Date->new->to_ymd, 'AWS4' . $self->access_key )
			)
		)
	);
}

sub _string_to_sign {
	my ( $self, $canonical_request, $date ) = @_;
	my $string_to_sign = $self->aws_auth_alg . "\n";
	$string_to_sign .= $date->to_string . "\n";
	$string_to_sign .= $date->to_ymd . '/' . $self->region . '/' . $self->aws_service . '/' . $self->aws_request . "\n";
	$string_to_sign .= sha256_hex( $canonical_request );
}

1;
__END__

=encoding utf-8

=head1 NAME

Mojo::S3Object - Mojo interface to S3 objects.

=head1 VERSION

0.02_5

=head1 SOURCE REPOSITORY

L<http://github.com/keenlinks/Mojo-S3Objects>

=head1 AUTHOR

Scott Kiehn E<lt>sk.keenlinks@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2018 - Scott Kiehn

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut
